acme/api/handler.go

```diff
@@ -184,10 +215,25 @@ func (h *Handler) GetChallenge(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	getLink := h.Auth.GetLink
-	w.Header().Add("Link", link(getLink(acme.AuthzLink, acme.URLSafeProvisionerName(prov), true, ch.GetAuthzID()), "up"))
-	w.Header().Set("Location", getLink(acme.ChallengeLink, acme.URLSafeProvisionerName(prov), true, ch.GetID()))
-	api.JSON(w, ch)
+	switch ch.Status {
+	case acme.StatusPending:
+		panic("validation attempt did not move challenge to the processing state")
+	// When a transient error occurs, the challenge will not be progressed to the `invalid` state.
+	// Add a Retry-After header to indicate that the client should check again in the future.
+	case acme.StatusProcessing:
+		w.Header().Add("Retry-After", ch.RetryAfter)
+		w.Header().Add("Cache-Control", "no-cache")
+		api.JSON(w, ch)
+	case acme.StatusInvalid:
+		api.JSON(w, ch)
+	case acme.StatusValid:
+		getLink := h.Auth.GetLink
+		w.Header().Add("Link", link(getLink(acme.AuthzLink, acme.URLSafeProvisionerName(prov), true, ch.GetAuthzID()), "up"))
+		w.Header().Set("Location", getLink(acme.ChallengeLink, acme.URLSafeProvisionerName(prov), true, ch.GetID()))
+		api.JSON(w, ch)
+	default:
+		panic("unexpected challenge state" + ch.Status)
```

Comment:
don't panic

Reply:
Does golang have asserts?

Reply:
Not that I know of. Our normal way of error handling is to let the errors bubble up through the stack. Your function returns an error and then the parent function gets to decide how it wants to handle the error.

Reply:
There are no asserts in the standard library, but you can create them. A C/C++ assert is basically 
```
func assert(expession bool) {
    if (expression) {
        return
    }
    panic("assertion failed")
}
```

On HTTP servers panics are recovered at top level (you can also recover them before), and they cause a 500 error, but I agree with Max and I think is better to return an error, I think our errors will also be formatted as json and might be more meaningfull than just a panic.

Reply:
Right, I do that in many places. But there are a few different classes of errors. Sometimes you want to indicate that there's something horribly wrong like a logical bug in the code vs a spurious read/write error. What I'm trying to do here is signal a logical error in the code. This is an error for developer eyes not user eyes. What can a user do with a "500: Internal server error the program is not operating the way it was intended" sort of error?

Reply:
You can create your own error type, if you like, and handle it in any fine-grained way you like. I think it's better to use the existing error handling infrastructure, one that we have firm control over. 

As to what the user sees, errors from the CA are printed in the CA logs with the full stack trace. Whatever message you put in the error will appear as part of that trace. If you have a descriptive and useful message, that's what will appear.

Reply:
But on an HTTP server, the user will see a 500 error, you can panic before the server initializes, on `init` for example.